provider/adapters/src/fallback/adapter.rs

```diff
@@ -3,25 +3,186 @@
+    /// let req = DataRequest {
+    ///     options: locale!("de-CH").into(),
+    ///     metadata: Default::default(),
+    /// };
+    ///
+    /// // There is no "de-CH" data in the `HelloWorldProvider`
+    /// ResourceProvider::<HelloWorldV1Marker>::load_resource(&provider, &req)
+    ///     .expect_err("No data for de-CH");
+    ///
+    /// // `HelloWorldProvider` does not contain fallback data,
+    /// // but we can fetch it from `icu_testdata`, and then
+    /// // use it to create the fallbacking data provider.
+    /// let fallbacker = LocaleFallbacker::try_new(&icu_testdata::get_provider())
+    ///     .expect("Fallback data present");
+    /// let provider = LocaleFallbackProvider::new_with_fallbacker(provider, fallbacker);
+    ///
+    /// // Now we can load the "de-CH" data via fallback to "de".
+    /// let german_hello_world: DataPayload<HelloWorldV1Marker> = provider
+    ///     .load_resource(&req)
+    ///     .expect("Loading should succeed")
+    ///     .take_payload()
+    ///     .expect("Data should be present");
+    ///
+    /// assert_eq!("Hallo Welt", german_hello_world.get().message);
+    /// ```
+    pub fn new_with_fallbacker(provider: P, fallbacker: LocaleFallbacker) -> Self {
+        Self {
+            inner: provider,
+            fallbacker,
+        }
+    }
+}
+
+impl<P> AnyProvider for LocaleFallbackProvider<P>
+where
+    P: AnyProvider,
+{
+    fn load_any(&self, key: ResourceKey, base_req: &DataRequest) -> Result<AnyResponse, DataError> {
+        let key_fallbacker = self.fallbacker.for_key(key);
+        let mut fallback_iterator = key_fallbacker.fallback_for(base_req.clone());
+        while !fallback_iterator.get().options.is_empty() {
+            let result = self.inner.load_any(key, fallback_iterator.get());
+            if !result_is_err_missing_resource_options(&result) {
+                return result;
+            }
+            fallback_iterator.step();
+        }
+        Err(DataErrorKind::MissingResourceOptions.with_req(key, base_req))
+    }
+}
 
-impl<P> BufferProvider for LocaleFallbackAdapter<P>
+impl<P> BufferProvider for LocaleFallbackProvider<P>
 where
-    P: LocaleFallbackProvider + BufferProvider,
+    P: BufferProvider,
 {
     fn load_buffer(
         &self,
-        _key: ResourceKey,
-        _req: &DataRequest,
+        key: ResourceKey,
+        base_req: &DataRequest,
     ) -> Result<DataResponse<BufferMarker>, DataError> {
-        todo!()
+        let key_fallbacker = self.fallbacker.for_key(key);
+        let mut fallback_iterator = key_fallbacker.fallback_for(base_req.clone());
+        while !fallback_iterator.get().options.is_empty() {
+            let result = self.inner.load_buffer(key, fallback_iterator.get());
+            if !result_is_err_missing_resource_options(&result) {
+                return result;
+            }
+            fallback_iterator.step();
+        }
+        Err(DataErrorKind::MissingResourceOptions.with_req(key, base_req))
```

Comment:
This is the same code as in `load_any`, `load_resource`, and `load_payload`. Can we deduplicate this?

Maybe `key_fallbacker.fallback_for(base_req.clone(), |req| self.inner.load_buffer(key, req))`?

Reply:
Done